gaps/gap-16_golem_compose/gap-16_golem_compose.md

```diff
@@ -23,55 +23,76 @@ The main motivation for this proposal is providing support for development of di
+An application hosted on Golem can be described by a set of data structures, which specify and represent all components of the applciationwhich are relevant from Golem's point of view. The data structures are an Object Model of an application, which has following attributes:
+- Object Model has a schema - it consists of object types (for example a `payload`, a `service`, a `network`), and each object type has a set of fields/attributes (eg. a `payload` includes a `runtime`, `capabilities` and runtime-specific `params`)
+- Objects have relationships (eg a `services` is specified by a `runtime`, and may be part of a `network`)
+- Object Model can be maintained in memory of an _engine_
+- Object Model is dynamically updated as the state of represented objects changes (eg. as a result of provisioning resources on Golem network)
+
+The _engine_ which is a host for an Object Model is responsible for:
+- ingesting an initial application specification (a *descriptor*)
+- provisioning Golem resources as per the _descriptor_
+- maintaining the Object Model state (eg. as resources are provisioned)
+
+
+[TODO: link to GAOM schema description] 
+Once the format of the descriptor YAML is finalized, its schema can be published to https://www.schemastore.org/json/.
+This way, the YAML language server will provide support for schema validation and completion in IDEs and editors.
+
 ### Configuration descriptor file
-Here's an example of such a descriptor file:
+An application descriptor specifies initial Golem Application Object Model. It must include all data required to provision Golem resources required by the application.
+The proposed format is YAML. 
+
+Here's an example of such a descriptor:
 ```
 version: "0.0.1"
 
-auth:
-  address: "http://rest-api:6000"
-  app_key: "$YAGNA_APP_KEY"
-  gsb_url: "tcp://gsb-url:6666"
-
-payment:
-  budget: 10  # GLM
-  driver: "polygon"
-  network: "mainnet"
-
 payloads:
-  nginx:
-    runtime: "vm" # runtimes should have their own capabilities assigned
+  web-server:
+    runtime: vm
+    constraints:
+      - "golem.inf.cpu.cores=2"
+      - "golem.inf.mem.gib>=4"
+    capabilities:
+      - "vpn"
     params:
-      image: "image-hash"
-      repo-url: "http://girepo.dev.golem.network:8000"
+      image_hash: "c37c1364f637c199fe710ca62241ff486db92c875b786814c6030aa1"
+
+  db-server:
+    runtime: vm
     constraints:
-      "golem.inf.cpu.cores": 2
-      "golem.inf.mem.gib": 4
-      "golem.inf.storage.gib": 10
+      - "golem.inf.storage.gib>10"
     capabilities:
       - "vpn"
+    params:
+      image_hash: "85021afecf51687ecae8bdc21e10f3b11b82d2e3b169ba44e177340c"
 
 networks:
   default:
-    ip: "192.168.0.0/24"
-
-nodes:
-  reverse-proxy:
-    payload: "nginx"
-    args:
-      deploy: # dict, optional
-      start: # list, optional
-    networks:
-      - "name": "default"
-        "address": "192.168.0.2"  # optional
-    environment:
-      - "SOME_VARIABLE=1234"
+    mask: "192.168.0.1/24"
```

Comment:
I'd stay with `ip` since this is the _address_ of the network, not its netmask - the netmask is an optional component here

Reply:
btw, @stranger80 ^ ?

Reply:
@stranger80 a friendly reminder, please rename this to `ip`, both in the examples and in the schema

Reply:
@shadeofblue done, although I still feel uneasy about this - as the field seems to specify not a single address, but rather IP range, correct?

Reply:
@stranger80 yes - it's not an address of a single location on a network but it's still an address and not a netmask ... if we feel uncomfortable with `ip`, maybe we can call it an `address`?